homework/list.js

```diff
@@ -1,79 +1,83 @@
+    this._listNoteContent = [];
+    this._wrapper = document.querySelector('#note-list-wrapper');
+}
+ToDoList.prototype = {
+    start: function() {
+	this.fetchList(function(data) {
+	    this.updateList(data);
+	    this.drawList();
+	    this.preloadFirstNote();
+	}.bind(this));
+	window.addEventListener('click', function(event) {
+	    this.onNoteOpen(event);
+	}.bind(this));
+    },
 
-(function() {
+    onNoteOpen: function(event) {
+	if (event.target.classList.contains('note-title')) {
+	    var id = event.target.dataset.noteId;
+	    var content = this._listNoteContent[id];
+	    window.dispatchEvent(new CustomEvent('note-open',
+						 { detail: content }));
+	};
+    },
 
-  var _listNoteContent = [];
-  var _wrapper = document.querySelector('#note-list-wrapper');
+    preloadFirstNote: function() {
+	if (this._listNoteContent.length !== 0) {
+	    var content = this._listNoteContent[0];
+	    window.dispatchEvent(new CustomEvent('note-open',
+						 { detail: content }));
+	}
+    },
 
-  function start() {
-    fetchList(function(data) {
-      updateList(data);
-      drawList();
-      preloadFirstNote();
-    });
-    window.addEventListener('click', function(event) {
-      onNoteOpen(event);
-    });
-  }
+    updateList: function(list) {
+	this._listNoteContent = list;
+    },
 
-  function onNoteOpen(event) {
-    if (event.target.classList.contains('note-title')) {
-      var id = event.target.dataset.noteId;
-      var content = _listNoteContent[id];
-      window.dispatchEvent(new CustomEvent('note-open',
-        { detail: content }));
-    };
-  }
+    drawList: function() {
+	var list = this._listNoteContent;
+	var ul = document.createElement('ul');
+	ul.id = 'note-title-list';
+	var buff = document.createDocumentFragment();
+	list.forEach(function(note, i) {
+	    var li = document.createElement('li');
+	    li.dataset.noteId = i;
+	    li.classList.add('note-title');
+	    li.textContent = note.title;
+	    // Note: buff is captured, so we now have a
+	    // little closure naturally.
+	    buff.appendChild(li);
+	});
+	ul.appendChild(buff);
+	this._wrapper.appendChild(ul);
+    },
 
-  function preloadFirstNote() {
-    if (_listNoteContent.length !== 0) {
-      var content = _listNoteContent[0];
-      window.dispatchEvent(new CustomEvent('note-open',
-        { detail: content }));
+    fetchList: function(afterFetch) {
+	var xhr = new XMLHttpRequest();
+	new Promise(function(resolve, reject){
+	    xhr.open('GET', '/demo-list-notes.json', true);
+	    xhr.responseType = 'json';
+	    xhr.onreadystatechange = function(e) {
+		// Watch out: we have a mysterious unknown 'this'.
+		if (this.readyState === 4 && this.status === 200) {
+		    // The flow ends here.
+		    resolve();
+		} else if (this.status !== 200 ){
+		    // Ignore error in this case.
+		    reject();
+		}
+	    };
+	    xhr.send();
+	}).then(function(){
+	    afterFetch(xhr.response);
+	}).catch(function(){
+	    console.log("fetchList failed.")
```

Comment:
It's a serious issue in real code that to catch an Error but not re-throw it. Because the following chain can't now be interrupted by the error, so it could devastate your debugging strategy. If you want to log it, you should alway re-throw the error after that.